File: rigour/names/prefix.py
import re
from functools import cache
from typing import List

from rigour._core import (
    obj_name_prefixes_list,
    org_name_prefixes_list,
    person_name_prefixes_list,
)


def _build_prefix_regex(prefixes: List[str]) -> re.Pattern[str]:
    escaped = "|".join(re.escape(p) for p in prefixes)
    return re.compile(r"^\W*((%s)\.?\s+)*" % escaped, re.I | re.U)


@cache
def _person_prefix_regex() -> re.Pattern[str]:
    return _build_prefix_regex(person_name_prefixes_list())


@cache
def _org_prefix_regex() -> re.Pattern[str]:
    return _build_prefix_regex(org_name_prefixes_list())


@cache
def _obj_prefix_regex() -> re.Pattern[str]:
    return _build_prefix_regex(obj_name_prefixes_list())


def remove_person_prefixes(name: str) -> str:
    """Strip honorific prefixes from the head of a person name.

    Drops `"Mr."`, `"Mrs."`, `"Dr."`, `"Lady"`, etc. so honorifics
    don't contaminate part alignment in matching or token-bag
    comparison. The list is data-driven from
    `resources/names/stopwords.yml::PERSON_NAME_PREFIXES`,
    surfaced via `rigour._core.person_name_prefixes_list`.

    Args:
        name: A person name string.

    Returns:
        The name with any leading honorific(s) removed. Idempotent
        for inputs that don't start with a known prefix.
    """
    return _person_prefix_regex().sub("", name)


def remove_org_prefixes(name: str) -> str:
    """Strip article-like prefixes from the head of an organisation name.

    Drops `"The"`, etc. so `"The Charitable Trust"` →
    `"Charitable Trust"` doesn't penalise the shorter variant
    when matching. Driven by
    `resources/names/stopwords.yml::ORG_NAME_PREFIXES`.

    Args:
        name: An organisation name string.

    Returns:
        The name with any leading article-prefix(es) removed.
    """
    return _org_prefix_regex().sub("", name)


def remove_obj_prefixes(name: str) -> str:
    """Strip vessel-class and generic-article prefixes from the
    head of an object name.

    Drops `"M/V"`, `"SS"`, `"The"`, etc. so `"M/V Oceanic"` →
    `"Oceanic"` doesn't penalise the shorter variant when
    matching vessels, vehicles, or aircraft. Driven by
    `resources/names/stopwords.yml::OBJ_NAME_PREFIXES`.

    Args:
        name: An object (vessel / vehicle / aircraft) name string.

    Returns:
        The name with any leading prefix(es) removed.
    """
    return _obj_prefix_regex().sub("", name)

File: rust/src/names/prefix.rs
//! Leading-prefix removers for person, organisation, and object
//! names — strip honorifics, articles, and vessel-class markers from
//! the head of a name before matching or indexing.

use std::sync::LazyLock;

use regex::Regex;

use crate::names::stopwords::{
    obj_name_prefixes_list, org_name_prefixes_list, person_name_prefixes_list,
};

fn build_prefix_regex(prefixes: &[String]) -> Regex {
    // Escape alternatives, join with `|`, wrap in the Python-compatible
    // anchored shape: `^\W*((alt|alt|...)\.?\s+)*`. Matches an optional
    // leading punctuation run, then zero-or-more prefix tokens each
    // followed by an optional `.` and required whitespace.
    let joined = prefixes
        .iter()
        .map(|p| regex::escape(p))
        .collect::<Vec<_>>()
        .join("|");
    let pattern = format!(r"(?i)^\W*(({})\.?\s+)*", joined);
    Regex::new(&pattern).expect("prefix regex compiles")
}

static PERSON_PREFIX_RE: LazyLock<Regex> =
    LazyLock::new(|| build_prefix_regex(&person_name_prefixes_list()));

static ORG_PREFIX_RE: LazyLock<Regex> =
    LazyLock::new(|| build_prefix_regex(&org_name_prefixes_list()));

static OBJ_PREFIX_RE: LazyLock<Regex> =
    LazyLock::new(|| build_prefix_regex(&obj_name_prefixes_list()));

/// Strip honorific prefixes ("Mr.", "Mrs.", "Dr.", "Lady", …) from
/// the head of a person name. Called before analysing or matching to
/// stop honorifics from contaminating part alignment.
pub fn remove_person_prefixes(name: &str) -> String {
    PERSON_PREFIX_RE.replace(name, "").into_owned()
}

/// Strip article-like prefixes ("The", …) from the head of an
/// organisation name. Drops "The Charitable Trust" → "Charitable
/// Trust" so matching doesn't penalise the shorter variant.
pub fn remove_org_prefixes(name: &str) -> String {
    ORG_PREFIX_RE.replace(name, "").into_owned()
}

/// Strip vessel-class markers ("M/V", "SS", …) and generic articles
/// from the head of an object name. `"M/V Oceanic"` → `"Oceanic"`.
pub fn remove_obj_prefixes(name: &str) -> String {
    OBJ_PREFIX_RE.replace(name, "").into_owned()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn person_prefix_dropped() {
        assert_eq!(remove_person_prefixes("Mr. John Smith"), "John Smith");
        assert_eq!(remove_person_prefixes("Mrs Mary Doe"), "Mary Doe");
    }

    #[test]
    fn person_prefix_case_insensitive() {
        assert_eq!(remove_person_prefixes("MR. John Smith"), "John Smith");
        assert_eq!(remove_person_prefixes("dr John Smith"), "John Smith");
    }

    #[test]
    fn person_prefix_no_match_is_identity() {
        assert_eq!(remove_person_prefixes("John Smith"), "John Smith");
    }

    #[test]
    fn org_prefix_dropped() {
        assert_eq!(remove_org_prefixes("The Acme Corp"), "Acme Corp");
    }

    #[test]
    fn org_prefix_no_match_is_identity() {
        assert_eq!(remove_org_prefixes("Acme Corp"), "Acme Corp");
    }

    #[test]
    fn obj_prefix_dropped() {
        assert_eq!(remove_obj_prefixes("M/V Oceanic"), "Oceanic");
    }
}
